### ISSUE TITLE
Type analyzer exhibits poor performance in nested calls

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1303]
This is related to a bug reported in the pylance repo: https://github.com/microsoft/pylance-release/issues/590#issuecomment-748167003

I've narrowed down the problem to the following expression, which is taking over 3000ms to analyze! Still need to figure out why.

```python
list(
    map(
        int,
        list(
            "".join(
                map(lambda x: x, list(map(chr, list(b""))))
            )
        ),
    )
)
```

**Comment by erictraut** [comment]
I've added additional type caching to cover the case of deeply-nested expressions of this sort.

**Comment by erictraut** [comment]
The above code previously required about 3000ms, and it now completes in about 80ms.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.97, which I just published. It will also be included in the next release of pylance.
